### PR TITLE
ExternalTaskSensor respects soft_fail if the external task enters a failed_state

### LINKED ISSUE #19754: ExternalTaskSensor should skip if soft_fail=True and external task in one of the failed_states
### Apache Airflow version

2.1.4

### Operating System

Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### What happened

I ran into a scenario where if I use an `ExternalTaskSensor` and set it to `soft_fail=True` but also set it to `failed_states=['skipped']` I would expect if the external task skipped then to mark this sensor as skipped, however for the `failed_states` check in the poke method if it is in one of those states it will explicitly fail with an `AirflowException`.

Wouldn't it make more sense to skip because of the `soft_fail`?

### What you expected to happen

The `ExternalTaskSensor` task should skip

### How to reproduce

1. Add a DAG with a task that is set to skip, such as this `BashOperator` task set to skip taken from https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/bash.html#skipping:
```
this_will_skip = BashOperator(
    task_id='this_will_skip',
    bash_command='echo "hello world"; exit 99;',
    dag=dag,
)
```
2. Add a second DAG with an `ExternalTaskSensor`
3. Set that sensor to have `external_dag_id` be the other DAG and `external_task_id` be the skipped task in that other DAG and `failed_states=['skipped']` and `soft_fail=True`
4. The `ExternalTaskSensor` fails instead of skips

### Anything else

I don't know what is desirable for most Airflow users:

1. To have `soft_fail` to only cause skips if the sensor times out? (like it seems to currently do)
2. To have `ExternalTaskSensor` with `soft_fail` to skip any time it would otherwise fail, such as the external task being in one of the `failed_states`?
3. To have some other way for the `ExternalTaskSensor` to skip if the external task skipped?

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by argibbs**
# Problem

You have an external task which may "skip". When that happens, you want the external task sensor to also skip. This PR enables that behaviour.

# Solution

The "right" way to do this is:
1. Set `failed_states` to `[State.SKIPPED]` (by default it is `None`)
2. Set `soft_fail=True`

With the above settings, when the target task enters the SKIPPED state, the ExternalTaskSensor will see this as a failed state and immediately react to the failure. Because of the `soft_fail` flag, it should not mark as failed, but instead as skipped.

# Effect of change

Prior to this change, the `ExternalTaskSensor` ignored the `soft_fail` setting, and always marked it as an error. With this change in place, the `ExternalTaskSensor` will instead mark the task as skipped, as desired.

# Other comments

`soft_fail` would already kick in if the ExternalTaskSensors timed out (i.e. if the external task did not enter an allowed_state or failed_state in the time. That means a possible workaround for the above is to set a timeout on the ExternalTaskSensor which you believe is long enough that the external task should have completed.

However (in my experience at least) ExternalTaskSensors are often run without timeouts. This is because if the target task hangs on upstream dependencies, having the external sensor timeout means cascading failures throughout the system which normally means more cleanup. It's usually easier to have the ExternalTaskSensors not timeout and thus only fail if there's an actual problem.

Additionally, soft_fail + execution timeouts is not perfect. If the task skips you may want the skip to propagate immediately; under the workaround, the skip would only kick in after the timeout. And of course, there are obvious goldilocks issues with finessing the timeout to be not too short nor too long.

# Related Issues

closes: #19754

### CODE REVIEW

**Comment by ferruzzi** [comment]
Looks reasonable to me.

**Review by potiuk — CHANGES_REQUESTED** [review]
This looks cool, however this is a change in behaviour which should be documented. It's not a breaking change I think, I'd rather lean towards a bugfix (It should have behaved like that already). I think though this one is enough of a change to warrant a newsfragment (of bugfix type) https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#step-4-prepare-pr

Can you please add it (and also add information about this behaviour in the docstring of the External Task Sensor @argibbs ?

@jedcunningham @ephraimbuddy - this is not a "critical enough" to make into 2.3.1 so I'd leave it up to you whether to cherry-pick / add it to 2.3.1 (though it should be small and localized enought to not be a problem to cherry-pick if we alll agree this is a "bugfix".

**Comment by argibbs** [comment]
Sorry, missed the comment, will crack on with the requested changes ASAP.

File: airflow/sensors/external_task.py
Comment (josh-fell) [inline airflow/sensors/external_task.py]:
```suggestion
                        f'The external DAG {self.external_dag_id} failed. Skipping due to soft_fail.'
```
Cleaning up a little implicit string concatenation.

File: newsfragments/23647.bugfix.rst
Comment (josh-fell) [inline newsfragments/23647.bugfix.rst]:
```suggestion
``ExternalTaskSensor`` now supports the ``soft_fail`` flag to skip if external task or DAG enters a failed state.
```
Trivial nit.

File: tests/sensors/test_external_task_sensor.py
Comment (josh-fell) [inline tests/sensors/test_external_task_sensor.py]:
```suggestion
            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID},
```

File: tests/sensors/test_external_task_sensor.py
Comment (josh-fell) [inline tests/sensors/test_external_task_sensor.py]:
```suggestion
            params={'dag_id': TEST_DAG_ID, 'task_id': TEST_TASK_ID},
```

File: tests/sensors/test_external_task_sensor.py
Comment (josh-fell) [inline tests/sensors/test_external_task_sensor.py]:
```suggestion
            allowed_states=[State.Failed],
            failed_states=[State.Success],
```
WDYT about using the class attr rather than the harcoded value?

**Review by josh-fell — COMMENTED** [review]
This is very cool! I love all the extra documentation in the operator too; very helpful.  Some nits and suggestions so other static checks are happier before releasing all of the other ones in CI. And it's a good idea to [run the static checks locally](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst) too if you can.

File: airflow/sensors/external_task.py
Comment (argibbs) [reply airflow/sensors/external_task.py]: Ah sorry - this is me running black and not properly vetting the result. Will fix.

File: tests/sensors/test_external_task_sensor.py
Comment (argibbs) [reply tests/sensors/test_external_task_sensor.py]: This is a black thing ... I don't mind removing, but just to warn you, if someone re-runs black on the file, it'll come back

File: tests/sensors/test_external_task_sensor.py
Comment (argibbs) [reply tests/sensors/test_external_task_sensor.py]: Ditto "I-blame-black"

File: tests/sensors/test_external_task_sensor.py
Comment (argibbs) [reply tests/sensors/test_external_task_sensor.py]: Good call.

File: tests/sensors/test_external_task_sensor.py
Comment (josh-fell) [reply tests/sensors/test_external_task_sensor.py]: Of course. Maybe CI black likes the one line, maybe it doesn't. Let's see! Totally fine if it comes back. If black is fine with the one line, I didn't want you to get dinged for it.

**Comment by argibbs** [comment]
Ok, I've run black again, weirdly black via breeze seems to do something different to black run externally 🤷 .

Anyhoo, `breeze static-checks -t black` passes locally for me.

**Comment by argibbs** [comment]
Yeah, I ran it from the root of the checkout... I dunno what I did wrong. I'm really hoping this is it now though. 😄 

It's been a learning curve but I'll hopefully walk into fewer rakes if (when?) I pick up another PR.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Tests are failing :)

**Comment by josh-fell** [comment]
> Tests are failing :)

Yeah, that's my fault. I provided an incorrect suggestion. @argibbs the state values should be `State.FAILED` and `State.SUCCESS` in your test.

**Comment by argibbs** [comment]
OK, <counts on fingers> 4th time's the charm

**Comment by argibbs** [comment]
Just checking but I assume the helm tests timing out are not a blocker? Do they need to be rerun, or do we just not mind given the unrelated nature of the change, or \<insert option C here\>?

**Comment by argibbs** [comment]
@potiuk @josh-fell just checking I don't have to do anything here...

I see there are some fixes for flakey tests hitting master; do you need me to rebase?

**Comment by potiuk** [comment]
> @potiuk @josh-fell just checking I don't have to do anything here...
> 
> I see there are some fixes for flakey tests hitting master; do you need me to rebase?

It's always good idea no matter what/

**Comment by argibbs** [comment]
> It's always good idea no matter what

Done. I'm not at my laptop, so have used github's web ui to rebase to main ... I _think_ it's done the right thing, o frabjous day.

**Review by josh-fell — APPROVED** [review]
Pending "fun with tests". Great addition!

**Comment by argibbs** [comment]
Fair beans, ty for the info. What's the best place to track the progress of the fixes? Should I just lurk on the devel slack for a Eureka post, or is there an issue or something I can watch? (Don't want to be rebasing and asking for reruns unnecessarily)

**Comment by potiuk** [comment]
> Fair beans, ty for the info. What's the best place to track the progress of the fixes? Should I just lurk on the devel slack for a Eureka post, or is there an issue or something I can watch? (Don't want to be rebasing and asking for reruns unnecessarily)

Just rebase often and early. No problem with that. You have ONE PR to take care about the maintainers about 30-50 things to look at every day (see my talk here https://www.youtube.com/watch?v=G6VjYvKr2wQ&t=1s - to get the right perspective.

Be persistent. rebase when you feel like it and see your PR is green. It's you who want to merge it, and just make sure things are gree. I rebase it now - but I leave it up to you if it is fixed and green (you can continue rebasing when you see it's not}. 

Ping us please if you see it's green and ready for review.

**Comment by argibbs** [comment]
> Be persistent.

Yeah, ok. I've no problem rebasing daily, but unless something's changed, as a first time committer I don't have the ability to run the workflows. I *was* trying to avoid being "that guy" who pops up every day asking about a low prio MR, precisely *because* you guys have so much to deal with. 😄 

Will see how it goes.

Ty!

**Comment by potiuk** [comment]
The thing is that if nothing changes, we are not notified about your new push with fixes. Probably every day is too often (You need to consider people\s time availabiliyt etc. ). Apache Software Foundation has this very good rule of giving people minimum 72 Hours for important stuff - like voting and participating in discussions, this is the right baseline. And of course everyone's mileage vary.  

The important thing is - that if you are fresh committer to GitHub and the project - any push of yours will require from us to approve your workflow (this is security protection against bots mining bitcoins by pull requests) so until you get your first PR merged, you need to ping in the PR to approve the workflow.

**Comment by argibbs** [comment]
Yup, all makes sense (esp the backdoor bitcoin mining thing sadly; this is why we can't have nice things). Love the app, trying to do my part and not get in the way 😄 

I'll check in every +/-3 days  (unless this latest attempt works!)

Thanks for the help!

**Comment by argibbs** [comment]
@potiuk @josh-fell I've rebased onto master, and resolved a minor conflict against main arising from tweaks to the file imports.

If you could approve the workflow, I'd like to roll the dice again :smile:

**Comment by argibbs** [comment]
Static checks passing locally now.

**Review by potiuk — APPROVED** [review]

**Comment by argibbs** [comment]
@potiuk @josh-fell w00t, it's passed. Can either of you merge it for me? I (understandably) don't have write access.

**Comment by argibbs** [comment]
Fantastic, thanks for all the help!

**Pull request merged**
